File: AppSets-Android/appsets/src/main/kotlin/xcj/app/appsets/im/ImObj.kt
package xcj.app.appsets.im

import xcj.app.appsets.server.model.Application
import xcj.app.appsets.server.model.GroupInfo
import xcj.app.appsets.server.model.UserInfo
import xcj.app.appsets.usecase.RelationsUseCase
import xcj.app.starter.android.util.PurpleLogger

sealed class ImObj(
    val bio: Bio,
    var isRelated: Boolean = false
) {
    val id: String
        get() = bio.id
    val name: String
        get() = bio.name ?: id
    val avatarUrl: Any?
        get() = bio.bioUrl

    class ImSingle(
        bio: Bio,
        val userRoles: String? = null
    ) : ImObj(bio, RelationsUseCase.getInstance().hasUserRelated(bio.id))

    class ImGroup(
        bio: Bio,
        val bios: List<Bio>? = null
    ) : ImObj(bio, RelationsUseCase.getInstance().hasGroupRelated(bio.id)) {

    }

    companion object {

        private const val TAG = "ImObj"

        fun fromBio(bio: Bio): ImObj {
            val imObj = when (bio) {
                is UserInfo -> {
                    ImSingle(bio)
                }

                is GroupInfo -> {
                    ImGroup(bio, bio.userInfoList)
                }

                is Application -> {
                    ImGroup(bio)
                }

                else -> ImSingle(bio)
            }
            PurpleLogger.current.d(TAG, "fromBio, bio:$bio, imgObj:${ImObj}")
            return imObj
        }
    }

}

File: AppSets-iOS/AppSets/com/app/appsets/im/ImObj.swift
//
//  ImObj.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/9.
//

import Foundation

protocol ImObj {
    
    var bio: any Bio { get set }
    
    var isRelated:Bool { get }
    
    var id:String { get }
    
    var name:String { get }
    
    var avatarUrl:String? { get }
}

struct ImObjStatic {
    
    static func fromBio(_ bio: any Bio) -> any ImObj {
        switch bio{
        case is UserInfo:
            return ImSingle(bio)
            
        case is GroupInfo:
            return ImGroup(bio)
            
        case is Application:
            return ImGroup(bio)
            
        default:
            return ImSingle(bio)
        }
    }
    
}


class ImSingle : ImObj {
    
    var bio: any Bio
    var isRelated: Bool
    
    var id: String {
        return bio.id
    }
    
    var name: String {
        return bio.name ?? bio.id
    }
    
    var avatarUrl: String? {
        return bio.bioUrl
    }
    
    let userRoles: String?
    
    init(_ bio: any Bio, userRoles:String? = nil) {
        self.bio = bio
        self.userRoles = userRoles
        self.isRelated = RelationsUseCase.Instance.hasUserRelated(bio.id)
    }
}

class ImGroup : ImObj {
    
    var bio: any Bio
    
    var isRelated: Bool
    
    var id: String {
        return bio.id
    }
    
    var bios:[any Bio]? = nil
    
    var name: String {
        return bio.name ?? bio.id
    }
    
    var avatarUrl: String? {
        return bio.bioUrl
    }
    
    init(_ bio: any Bio) {
        self.bio = bio
        self.isRelated = RelationsUseCase.Instance.hasGroupRelated(bio.id)
    }
    
}
